### PR TITLE
[p2p/utils] Implement `Requester` (Using `PrioritySet`)

### PR DESCRIPTION
**Pull request by patrick-ogrady**
Related: #180 

`Requester` was implemented hastily in `commonware-consensus::authority/backfiller`. This is a cleaner and well-tested implementation of the same intertwined logic: https://github.com/commonwarexyz/monorepo/blob/bb8497a131e79ba04dfa2735c674860448bd2568/consensus/src/authority/actors/voter/actor.rs.

### CODE REVIEW

**Comment by patrick-ogrady** [comment]
Make request opaque (can be a batch or not) and just be told what was not fulfilled when timeout?

File: utils/src/priority_set.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+impl<I: Ord + Hash + Clone, V: Ord + Clone> PartialOrd for Entry<I, V> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A set that offers fast, priority-ordered iteration over
+/// its elements.
+pub struct PrioritySet<I: Ord + Hash + Clone, V: Ord + Clone> {
```
Comment (patrick-ogrady) [inline utils/src/priority_set.rs:28]: The alternative approach to this (to reduce the copying on `put`) would be to use an indexed heap.

File: utils/src/priority_set.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+impl<I: Ord + Hash + Clone, V: Ord + Copy> PartialOrd for Entry<I, V> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A set that offers fast, priority-ordered iteration over
+/// its elements.
+pub struct PrioritySet<I: Ord + Hash + Clone, P: Ord + Copy> {
+    entries: BTreeSet<Entry<I, P>>,
```
Comment (patrick-ogrady) [inline utils/src/priority_set.rs:30]: While it may look tempting to replace these two inner structures with an "indexed heap", the only object we wouldn't be copying would be `P` (which is expected to be very cheap to copy). I felt it wasn't worth the complexity.

File: utils/src/priority_set.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            self.entries.remove(&entry);
+        }
+
+        // Add any items not yet removed with the initial priority
+        for item in retained {
+            self.put(item.clone(), initial);
+        }
+    }
+
+    /// Iterate over all items in priority order.
```
Comment (BrendanChou) [inline utils/src/priority_set.rs]: indicate here (and maybe on the struct) the direction of the iteration (asc or desc)

File: utils/src/priority_set.rs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    /// Get the current priority of an item.
+    pub fn get(&self, item: &I) -> Option<P> {
+        self.keys.get(item).cloned()
+    }
+
+    /// Remove all previously inserted items not included in `keep`
+    /// and add any items not yet seen with a priority of `initial`.
+    pub fn reconcile(&mut self, keep: &[I], initial: P) {
```
Comment (BrendanChou) [inline utils/src/priority_set.rs]: nit: `default` rather than `initial` might be the correct term?

File: p2p/src/utils/requester.rs [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    }
+
+    /// Indicate which participants can be sent requests.
+    pub fn reconcile(&mut self, participants: &[PublicKey]) {
+        self.participants
+            .reconcile(participants, self.initial.as_millis());
+        self.rate_limiter.shrink_to_fit();
+    }
+
+    /// Once a participant is added to skip, it is never removed.
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: "never removed" is confusing. I read this as "never removed from the set" but it seems like you mean never used as part of the set?

File: p2p/src/utils/requester.rs [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+    timeout: Duration,
+
+    rate_limiter:
+        RateLimiter<PublicKey, HashMapStateStore<PublicKey>, E, NoOpMiddleware<E::Instant>>,
+    participants: PrioritySet<PublicKey, u128>,
+    excluded: HashSet<PublicKey>,
+
+    id: ID,
+    requests: HashMap<ID, (PublicKey, SystemTime, SystemTime)>,
+    deadlines: BTreeMap<SystemTime, BTreeSet<ID>>,
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: We don't expect a lot of overlapping deadlines (i.e. equal `SystemTime` values). Would it be simpler and more efficient to have a data structure of a single "layer"? That is, simply a sorted set of tuples `(SystemTime, ID)` which is sorted by time then id? Otherwise you need this sorted set of sets, which seems less efficient when we expect every inner set to have 1 value (rarely more).

File: p2p/src/utils/requester.rs [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+            return;
+        };
+        let performance = past / 2 + self.timeout.as_millis() / 2;
+        self.participants.put(participant.clone(), performance);
+    }
+
+    /// Cancel an outstanding request.
+    pub fn cancel(&mut self, id: ID) {
+        self.remove(id);
+    }
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: Why is both `cancel` and `remove` needed? Am I missing a trait that needs to be impl'ed?

Suggest removing (no pun intended) `cancel` if it's the same except with no return value

File: p2p/src/utils/requester.rs [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+    pub fn timeout(&mut self, id: ID) {
+        let Some((participant, _)) = self.remove(id) else {
+            return;
+        };
+
+        // Calculate new performance using exponential moving average
+        let Some(past) = self.participants.get(&participant) else {
+            return;
+        };
+        let performance = past / 2 + self.timeout.as_millis() / 2;
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: `a/2 + b/2` may be wrong compared to `(a+b)/2`, but does avoid overflow.

For example averaging 7 and 3 this way gives 4: `7/2 + 3/2 = 3 + 1 = 4`

This probably doesn't matter much for milliseconds but could also have some sort of `avg()` util function

File: p2p/src/utils/requester.rs [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+
+    /// Initial expected performance for new participants.
+    pub initial: Duration,
+
+    /// Timeout for requests.
+    pub timeout: Duration,
+}
+
+/// Send rate-limited requests to peers prioritized by performance.
+pub struct Requester<E: Clock + GClock, C: Scheme> {
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: would be cool to have small comments for each field (could be `//` instead of `///` to orient the reader)

File: p2p/src/utils/requester.rs
Comment (BrendanChou) [inline p2p/src/utils/requester.rs:1]: Is this meant to be probabilistic or round-robin in any way? Otherwise this can result to just sending the same message to the same `participant` over and over again.

For example take a high `initial` value. Some participant is chosen first and beats the initial value, lowering it. This participant will then be chosen for each of the next requests, forever.

One solution is to set the `initial` value to zero, but it could still theoretically get "stuck" in some configuration. I guess this might be fine since if the rate is low, it can just  go to the same participant over and over again with no issue, whereas if the rate is high, then it will cycle participants due to the rate limiter. Just wanted to make sure this is the expected behavior

File: p2p/src/utils/requester.rs (hunk 8 above)
Comment (BrendanChou) [reply p2p/src/utils/requester.rs]: Also, consider having a `learning_rate` in the config between 0 and 1, where:
`performance = past * (1-learning_rate) + timeout * learning_rate;`

File: p2p/src/utils/requester.rs
Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs:1]: This purposely saturates the bandwidth (bounded by rate limit) of the most performant peers (based on our latency observations). At some point, we could introduce some randomness here (i.e. shuffle iterator) periodically but felt that wasn't worth the complexity initially.

Will mull over a bit more to double-down or revisit.

File: p2p/src/utils/requester.rs (hunk 8 above)
Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs]: I avoided the learning_rate approach because I'd have to use floats (which aren't deterministic).

On the point above, this was lazy of me (and had the sum before). I'm going to update it to do a saturating_add before dividing by 2.

File: p2p/src/utils/requester.rs (hunk 7 above)
Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs]: `remove` is a helper to keep everything up to date internally whereas `cancel` is meant to be publicly accessible. Fair point around why have both...

File: p2p/src/utils/requester.rs (hunk 6 above)
Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs]: Yeah that's a good point. I'll use the same technique that I employ with `BTreeSet` in `priority_set`.

File: p2p/src/utils/requester.rs (hunk 5 above)
Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs]: I'm trying to convey that anything added here forever is "blocked" (even if it is dropped during `reconcile`). My thinking was that it was nice to keep across add->remove->add but that's probably overkill (and its worth retrying someone once they are removed there). I'll update this.

File: utils/src/priority_set.rs (hunk 4 above)
Comment (patrick-ogrady) [reply utils/src/priority_set.rs]: Someday I'll learn to speak english.

File: utils/src/priority_set.rs (hunk 3 above)
Comment (patrick-ogrady) [reply utils/src/priority_set.rs]: Fair. Will update.

File: p2p/src/utils/requester.rs
Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs:1]: Addressed this by allowing the caller to provide `shuffle` to `request()`. Also added an explanation of this decision.

File: p2p/src/utils/requester.rs [hunk 10]
```diff
@@ -35,23 +36,35 @@ pub struct Config<C: Scheme> {
...
     initial: Duration,
     timeout: Duration,
 
+    // Participants to exclude from requests
+    excluded: HashSet<PublicKey>,
+
+    // Rate limiter for participants
     rate_limiter:
         RateLimiter<PublicKey, HashMapStateStore<PublicKey>, E, NoOpMiddleware<E::Instant>>,
+    // Participants and their performance
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: nit: comment that lower is better (i.e. more performant) since it's a latency estimate (?)

File: p2p/src/utils/requester.rs [hunk 11]
```diff
@@ -78,29 +92,28 @@ impl<E: Clock + GClock, C: Scheme> Requester<E, C> {
         self.rate_limiter.shrink_to_fit();
     }
 
-    /// Once a participant is added to skip, it is never removed.
-    pub fn skip(&mut self, participant: PublicKey) {
+    /// Skip a participant for future requests.
+    ///
+    /// Participants added to this list will never be removed (even if dropped
+    /// during `reconcile`, in case they are re-added later).
+    pub fn block(&mut self, participant: PublicKey) {
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs:100]: Is `unblock` needed at all? Not sure the use of this function, primarily blocking byzantine actors?

File: p2p/src/utils/requester.rs [hunk 12]
```diff
@@ -78,29 +92,28 @@ impl<E: Clock + GClock, C: Scheme> Requester<E, C> {
...
+            participants.shuffle(&mut self.runtime);
         }
-        Some((participant, start))
-    }
 
-    /// Ask for a participant to handle a request.
-    pub fn request(&mut self) -> Option<(PublicKey, ID)> {
         // Look for a participant that can handle request
-        for (participant, _) in self.participants.iter() {
+        for (participant, _) in participants {
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: You may have realized this, but this is less efficient when not sorting. Is it possible somehow to keep it as an iterator? Something like:

```
let mut participants = self.participants.iter();
if shuffle {
  participants = participants.collect::<Vec<_>>().shuffle(&mut self.runtime).iter();
}

for (participant, _) in participants {
  // ...
```

**Review by BrendanChou — DISMISSED** [review]

File: p2p/src/utils/requester.rs (hunk 11 above)
Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs:100]: Once someone is blocked, we shouldn't reconsider interacting with them until restart.

This is needed to avoid spending resources verifying messages from people who typically send us invalid data (and to avoid waiting for them to respond).

File: p2p/src/utils/requester.rs (hunk 12 above)
Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs]: I took an initial stab at this (for the same reason you called out) but gave up because it wasn't clean.

Dug deeper this time and got it to compile with:
```
        // Shuffle participants
        let participants = self.participants.iter();
        let participants = if !shuffle {
            Box::new(participants) as Box<dyn std::iter::Iterator<Item = (&Bytes, &u128)>>
        } else {
            let mut participants = participants.collect::<Vec<_>>();
            participants.shuffle(&mut self.runtime);
            Box::new(participants.into_iter())
        };

```

Will think about it some more.

Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs]: You could avoid dynamic dispatch with something like this:
```
enum ParticipantIter<'a> {
    Unshuffled(std::slice::Iter<'a, (&Bytes, &u128)>),
    Shuffled(std::vec::IntoIter<(&'a Bytes, &'a u128)>),
}

impl<'a> Iterator for ParticipantIter<'a> {
    type Item = (&'a Bytes, &'a u128);

    fn next(&mut self) -> Option<Self::Item> {
        match self {
            ParticipantIter::Unshuffled(iter) => iter.next(),
            ParticipantIter::Shuffled(iter) => iter.next(),
        }
    }
}

pub fn request(&mut self, shuffle: bool) -> Option<(PublicKey, ID)> {
    let participants = if !shuffle {
        ParticipantIter::Unshuffled(self.participants.iter())
    } else {
        let mut shuffled = self.participants.iter().collect::<Vec<_>>();
        shuffled.shuffle(&mut self.runtime);
        ParticipantIter::Shuffled(shuffled.into_iter())
    };
    .....
```

Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs]: Oooooo found a clean way using: https://crates.io/crates/either

File: p2p/src/utils/requester.rs [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+            return;
+        };
+
+        // Calculate new performance using exponential moving average
+        let Some(past) = self.participants.get(&participant) else {
+            return;
+        };
+        let performance = past.saturating_add(self.timeout.as_millis()) / 2;
+        self.participants.put(participant.clone(), performance);
+    }
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: I wonder if these can actually just be combined into one function, reducing code duplication.

They are exactly the same behavior-wise in the case that the timeout happens exactly at the right time, and only slightly different otherwise. So would suggest removing the `timeout` function and adding instead `let elapsed = min(elapsed, self.timeout).as_millis();`

File: p2p/src/utils/requester.rs [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+            self.requests.insert(id, (participant.clone(), now));
+            let deadline = now.checked_add(self.timeout).expect("time overflowed");
+            self.deadlines.put(id, deadline);
+            return Some((participant.clone(), id));
+        }
+        None
+    }
+
+    /// Resolve an outstanding request.
+    pub fn resolved(&mut self, id: ID) {
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: nit (grammar): the other functions are active verbs, so call this `resolve()` instead

File: p2p/src/utils/requester.rs [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+        // Remove request
+        let Some((participant, start)) = self.cancel(id) else {
+            return;
+        };
+
+        // Get elapsed time
+        let Ok(elapsed) = self.runtime.current().duration_since(start) else {
+            return;
```
Comment (BrendanChou) [inline p2p/src/utils/requester.rs]: tiny nit: consider `Ok(Duration::default())` instead of `return;` due to the non-monotonic possibility of `SystemTime`. This change would use `0` as the elapsed time and updates the participant's performance (rather than avoiding updating the participant's performance)

File: utils/src/priority_set.rs [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        self.keys.get(item).cloned()
+    }
+
+    /// Remove an item from the set.
+    pub fn remove(&mut self, item: &I) {
+        let Some(entry) = self.keys.remove(item).map(|priority| Entry {
+            item: item.clone(),
+            priority,
+        }) else {
+            return;
```
Comment (BrendanChou) [inline utils/src/priority_set.rs:78]: nit: looks like this line is untested, consider a simple test to remove an item that doesn't exist

File: p2p/src/utils/requester.rs (hunk 15 above)
Comment (patrick-ogrady) [reply p2p/src/utils/requester.rs]: Good catch!
